BaseStation/Utilities/ObservableDictionary.cs

```diff
@@ -0,0 +1,268 @@
+        //}
+
+        public IEnumerator<Entry> GetEnumerator()
+        {
+            return new DictionaryEnumerator(dictionary.GetEnumerator());
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        void ICollection<Entry>.Add(Entry item) => PutValue(item.Key, item.Value);
+
+        public void Clear() => dictionary.Clear();
+
+        // Hopefully this does what we want
+        public bool Contains(Entry item) => dictionary.Contains(item.Key); 
+
+        public void CopyTo(Entry[] array, int arrayIndex) => dictionary.CopyTo(array, arrayIndex);
+
+        public bool IsFixedSize {
+            get => false;
+        }
+
+        public bool Remove(Entry item)
+        {
+            if (dictionary.Contains(item.Key))
+            {
+                dictionary.Remove(item.Key);
+                return true;
+            } else
+            {
+                return false;
+            }            
+        }
+
+        public int IndexOf(Entry item)
+        {
+            int i = 0;
+            foreach (DictionaryEntry entry in dictionary)
+            {
+                if (entry.Key.Equals(item.Key))
+                {
+                    return i;
+                }
+                i++;
+            }
+            return -1;
+        }
+
+        public void Insert(int index, Entry item)
+        {
+            dictionary.Insert(index, item.Key, item.Value);
+        }
+
+        public void RemoveAt(int index)
+        {
+            dictionary.RemoveAt(index);
+        }
+
+        public int Add(object value)
+        {
+            return Add((Entry)value);
+        }
+
+        public bool Contains(object value)
+        {
+            return Contains((Entry)value);
+        }
+
+        public int IndexOf(object value)
+        {
+            return IndexOf((Entry)value);
+        }
+
+        public void Insert(int index, object value)
+        {
+            Insert(index, (Entry)value);
+        }
+
+        public void Remove(object value)
+        {
+            Remove((Entry)value);
+        }
+
+        public void CopyTo(Array array, int index)
+        {
+            CopyTo((Entry[])array, index); // The cast here might not work
+        }
+
+        public bool IsReadOnly => false;
+
+        public int Count => dictionary.Count;
+
+        public object SyncRoot => null; // Not a proper implementation
+
+        public bool IsSynchronized => false;
+
+        object IList.this[int index] { get => this[index]; set => this[index] = (Entry)value; }
+
+        public Entry this[int index] {
+            get {
+                var entry = dictionary.Cast<DictionaryEntry>().ElementAt(index);
+                return new Entry((K)entry.Key, (V)entry.Value);
+            }
+            set {
+                // A dumb way to do it, but it works and it preserves indexes when the key changes
+                dictionary.RemoveAt(index);
```

Comment:
This is a linear time insert. If dictionary is big, it will be very slow. If we don't expect dictionary to be big, then we can leave this. If we do want it to be big, it might be better to just change the key - maybe using reflection?